In [2]:
from tldextract import extract
import pandas as pd
import numpy as np
from tqdm import tqdm
from collections import Counter
import math

In [3]:
unlabeled = pd.read_csv('final_unlabeled.csv')
malicious = pd.read_csv('final_malicious.csv')

In [5]:
unlabeled.head()

Domain
0  http://-----6kcackgf6aecovid2bb2af.sds.miningb...
1  http://--antdotocoronavirus-jyb.com.ipaddress.com
2             http://--coronavirus-quarantne-o2b.com
3                  http://--coronavirusbelgi-prb.com
4                http://--coronavirusenespaa-30b.com

In [4]:
malicious_dom = malicious['URL'].values.tolist()
unlabeled_dom = unlabeled['Domain'].values.tolist()

In [5]:
from tldextract import extract
def check_tld(array):
    new_list = []
    for i in array:
        tsd, td, tsu = extract(i) #subdomain, domain, suffix
        if td:
            dom = td + '.' + tsu
        else:
            dom = tsu
        new_list.append(dom)
    print(len(new_list))
    print(len(np.unique(new_list)))
    return(np.unique(new_list))

In [9]:
l = check_tld(unlabeled_dom)

344332
121695


In [7]:
dict = {'Domain': list}

In [8]:
# pd.DataFrame(dict).to_csv('final_malicious_domains.csv', index=None, header = True)

## Original domains
### original number of domains in malicious : 61645 after dropping sub-domains 61400
### original number of domains in unlabelled : 344332 after dropping sub-domains 121695

In [21]:
def entropy(s):
    p, lns = Counter(s), float(len(s))
    return -sum(count/lns * math.log(count/lns,2) for count in p.values())

In [25]:
def avg_entropy(l):
    e = 0
    for i in l:
        e += entropy(i)
    return(e/(len(l)))

In [26]:
avg_entropy(list)

3.655807366522173

In [28]:
#drop the suffix
def check_td(array):
    new_list = []
    for i in array:
        tsd, td, tsu = extract(i) #subdomain, domain, suffix
        if td:
            dom = td
        else:
            dom = tsu
        new_list.append(dom)
    print(len(new_list))
    print(len(np.unique(new_list)))
    return(np.unique(new_list))

In [30]:
# check average entropy after dropping the suffix
avg_entropy(check_td(malicious_dom))

61645
44806


3.471170199230667

In [1]:
import ipaddress

def havingIP(url):
  try:
    ipaddress.ip_address(url)
    ip = 1
  except:
    ip = 0
  return ip


In [1]:
from sklearn import datasets

In [3]:
iris = datasets.load_iris

In [1]:
import dnstwist

In [26]:
fuzz = dnstwist.DomainFuzz("covid.com")
fuzz
#<dnstwist.DomainFuzz instance at 0xc300e0>
#fuzz.generate()
#fuzz.domains

In [27]:
fuzz.generate()

In [34]:
from tldextract import extract

In [38]:
_ , domain, _ = extract(fuzz.domains[999]['domain-name'])

In [40]:
keywords = ['corona', 'covid' , 'ncov', 'wuhan', 'ncov-19', 
            'virus', 'covid-19', 'covid19', 'sars', 'wuhanvirus',
            'novelvirus']

In [157]:
def generate_twists(keys):
    fuzz_doms = []
    fake_doms = []
    # generate twisted domains from the provided keywords
    
    for i in keys:
        i += '.com' # add a random suffix to pass through the fuzz generator
        fuzz = dnstwist.DomainFuzz(i)
        fuzz.generate()
        fuzz_doms += fuzz.domains[1:] # drop the first result in list as it is the original keyword
        
    # extract domain names from the generated list
    
    for i in fuzz_doms:
        _ , domain, _ = extract(i['domain-name'])
        fake_doms.append(domain)
    
    return fake_doms

def check_typo(twists, array):
    typo_doms = []
    for i in array:
        temp_mark = 0
        for j in twists:
            if (len(j) > 3) and (j in i):
                print(j, i)
                temp_mark += 1        
        if temp_mark > 0:
            typo_doms.append(1)
        else:
            typo_doms.append(0)
        
    print("Detected {:.2f}% of domains with typosquatting.".format((typo_doms.count(1)*100/len(typo_doms))))
    return typo_doms

In [158]:
fake_doms = generate_twists(keywords)

In [163]:
doms = ["na.com", "coronaa.com", "c0vid.com", "coronac.com"]

In [164]:
check_typo(fake_doms, doms)

coronaa coronaa.com
orona coronaa.com
coron coronaa.com
orona coronaa.com
rona coronaa.com
c0vid c0vid.com
coronac coronac.com
orona coronac.com
coron coronac.com
orona coronac.com
rona coronac.com
Detected 75.00% of domains with typosquatting.


[0, 1, 1, 1]

In [149]:
fake_doms[1000:]

['eoȓona',
 'ƈ0r0na',
 'coɼoņa',
 'ćoronä',
 'corȯnạ',
 'coɼoǹa',
 'cȯroꞑa',
 'çorora',
 'coŕoṇa',
 'corȯną',
 'coroꞑâ',
 'çoɍona',
 'coroñȧ',
 'c0rȯna',
 'cor0ņa',
 'cörönä',
 'ĉörona',
 'cơroną',
 'corọnå',
 'cọṙọna',
 'coṙona',
 'coɍoņa',
 'coṛona',
 'cọrọnă',
 'eorỏna',
 'coröną',
 'corỏnȧ',
 'cöröńa',
 'coṛoņa',
 'cơŕơna',
 'coṙơna',
 'corỏnä',
 'coroǹå',
 'coroňã',
 'cỏrỏnã',
 'corȯṉa',
 'ĉóróna',
 'coɾoǹa',
 'cöroná',
 'ċọrọna',
 'cörönǎ',
 'cỏronå',
 'cořonạ',
 'cỏrỏnâ',
 'eoroma',
 'cȯrỏna',
 'corơꞑa',
 'cọŗọna',
 'eoɽona',
 'cȯṟȯna',
 'cóɼóna',
 'coroňâ',
 'coroṇä',
 'ćorọna',
 'coȑonȧ',
 'corȯná',
 'coɍonâ',
 'coroǹá',
 'cơrơńa',
 'cóŕóna',
 'ƈoʀona',
 'coŕoña',
 'corönà',
 'cořoṇa',
 'cȯŗona',
 'c0ɽona',
 'coȓonă',
 'cơrơnâ',
 'coȓonà',
 'ċoȓona',
 'cořonã',
 'coʀonạ',
 'coṙonȧ',
 'coɼonå',
 'cỏŗỏna',
 'ĉoronà',
 'c0ř0na',
 'cọrọma',
 'ç0rona',
 'čoronɑ',
 'coȓoȓa',
 'coɼonà',
 'coɼonã',
 'coṙoꞑa',
 'čorȯna',
 'coɽơna',
 'coṙoṇa',
 'coɼơna',
 'cỏrỏnɑ',
 'cor0nă',
 'čoȓona',